src/app/proposals/shared/utils.ts

```diff
@@ -241,6 +237,57 @@ export function getProposalCategory(calldatasParsed: any[]): string {
   return hasWithdrawAction ? ProposalCategory.Grants : ProposalCategory.Activation
 }
 
+/**
+ * Extracts proposal category from parsed calldata and description
+ * @param calldatasParsed - Array of decoded calldata
+ * @param description - Original proposal description for milestone detection
+ * @returns ProposalCategory
+ */
+export function getProposalCategoryFromParsedData(
+  calldatasParsed: DecodedData[],
+  description: string,
+): ProposalCategory {
+  // Extract all decoded function names once
+  const decodedFunctionNames = calldatasParsed
+    .filter(data => data.type === 'decoded')
+    .map(data => data.functionName)
+
+  // Map function names to their categories
+  const functionCategoryMap = new Map<string, ProposalCategory>([
+    ['communityApproveBuilder', ProposalCategory.Activation],
+    ['whitelistBuilder', ProposalCategory.Activation],
+    ['dewhitelistBuilder', ProposalCategory.Deactivation],
+    ['removeWhitelistedBuilder', ProposalCategory.Deactivation],
+  ])
+
+  // Check for builder functions first
+  for (const functionName of decodedFunctionNames) {
+    const category = functionCategoryMap.get(functionName)
+    if (category) {
+      return category
+    }
+  }
+
+  // Then check for milestone (only if not a builder function)
+  const milestoneRegex = new RegExp(`${MILESTONE_SEPARATOR}(\\S+)`, 'i')
+  const milestoneMatch = description.match(milestoneRegex)
+
+  if (milestoneMatch) {
+    const milestoneNumber = milestoneMatch[1]
+    switch (milestoneNumber) {
+      case Milestones.MILESTONE_1:
+        return ProposalCategory.Milestone1
+      case Milestones.MILESTONE_2:
+        return ProposalCategory.Milestone2
+      case Milestones.MILESTONE_3:
+        return ProposalCategory.Milestone3
+    }
+  }
+
```

Comment:
maybe do search instead of manually comparing with each value?
```suggestion
const milestoneKey = Object.keys(Milestones).find(
  key => Milestones[key as keyof typeof Milestones] === match,
) as keyof typeof Milestones
const milestoneNumber = milestoneKey ? Milestones[milestoneKey] : undefined
```